File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <stdlib.h>

int ability(void){
    int sum = 0;
    int min = 0;
    for(int j=0; j<4; j++) {
        int throw = rand()%6 + 1;      
        sum += throw;
        min = (throw < min) ? throw : min;
    }
    return sum - min;    
}

int modifier(int score){
    int m = (score-10);
    return (m >= 0) ? m/2 : (m-1)/2;    
}


dnd_character_t make_dnd_character(void){
    dnd_character_t character;
    character.strength = ability();
   character.dexterity = ability();
   character.constitution = ability();
   character.intelligence = ability();
   character.wisdom = ability();
   character.charisma = ability();
   character.hitpoints = 10 + modifier(character.constitution);
   return character; 
}

File: solutions/dart/dnd-character/1/lib/dnd_character.dart
import 'dart:math';

class DndCharacter {
  late int strength;
  late int dexterity;
  late int constitution;
  late int intelligence;
  late int wisdom;
  late int charisma;
  late int hitpoints;
  
  static int modifier(int score) {
        int x = score-10;
        if (x<0) x--;
        return x~/2;    
   }

  static int ability(){
    Random rand = new Random();
    int sum = 0;
    int mi = 6;
    for(int i=0; i<4; i++){
      int r = rand.nextInt(6)+1;
      sum += r;
      mi = min(mi,r);
    }
    return sum-mi;
  }

  DndCharacter.create(){
    strength = ability();
    dexterity = ability();
    constitution = ability();
    intelligence = ability();
    wisdom = ability();
    charisma = ability();
    hitpoints = 10 + modifier(constitution);
  }
}
